README.md
# real-og

If you need a thing to generate social media [og:image](https://ogp.me) social
card images from the markdown pages on your static site, then this library may
help you piece together a script to do so.

## Usage

```js
const realOg = require('real-og')

// Async function, returns a promise
realOg('posts/my-post.md', 'img/my-post.png', {

  // Will be parsed as an EJS template and passed the following object:
  //   `data` - frontmatter of the markdown file
  //   `content.markdown` - Markdown below the frontmatter
  //   `content.html` - Compiled HTML of the Markdown below the frontmatter
  //   `summary.markdown` - Markdown up to the `summarySeparatory`
  //   `summary.html` - Compiled HTML up to the Markdown's `summarySeparatory`
  //   `globals.*` - Whatever you set as `globals` on this object
  template: 'template/social.html',

  // Optional options

  fileType: 'png', // jpg, gif, webp
  summarySeparator: '<!--more-->', // token that separates
  width: 1200,
  height: 630,
  globals: {
    /* Any additional stuff you want to send to the template. */
  },
  // Override the tailwind classes of elements generated by markdown
  markdownTailwindClasses: {
    p: 'mt-0 mb-6',
    h1: 'font-size-3xl',
  },
  // Override the style attribute for these markdown elements
  markdownStyleAttrs: {
    p: 'text-overflow: ellipsis',
  },
  // Remove these tags since Satori can't really handle inline text elements
  markdownUnwrapElements: ['a', 'em'],
  // Custom extensions for showdown's markdown processing.
  // See: https://github.com/showdownjs/showdown/wiki/Extensions
  markdownExtensions: [
    {
      type: 'output',
      regex: new RegExp(`<li([^>]*)>([^<]*)<\\/li>`, 'g'),
      replace: '<li$1>• $2</li>'
    }
  ],


  // At least one font is required by the satori package
  fonts: [
    {
      name: 'Source Sans Pro',
      data: fs.readFileSync(require.resolve('@fontsource/source-sans-pro/files/source-sans-pro-latin-400-normal.woff')),
      weight: 400,
      style: 'normal'
    }
  ]
})
```

The template can contain any of the subset of HTML and CSS supported by
[satori](https://github.com/vercel/satori), including
[Tailwind](https://tailwindcss.com) classes via a `tw` attribute.

# Acknowledgements

This package was started by adapting the meat of
[eleventy-plugin-og-imag](https://github.com/KiwiKilian/eleventy-plugin-og-image)
into a standalone package exposing its core functionality as an API that isn't
coupled to the separately-excellent [11ty](https://www.11ty.dev)

## Code of Conduct

This project follows Test Double's [code of
conduct](https://testdouble.com/code-of-conduct) for all community interactions,
including (but not limited to) one-on-one communications, public posts/comments,
code reviews, pull requests, and GitHub issues. If violations occur, Test Double
will take any action they deem appropriate for the infraction, up to and
including blocking a user from the organization's repositories.


CHANGELOG.md
# 0.0.8

- Add undocumented `measureHeight` function. YMMV

# 0.0.7

- Remove errant log statement

# 0.0.6

- Add `markdownExtensions` and `markdownStyleAttrs` options

# 0.0.5

- Drop `sartori-html` for much faster `html-react-parser`

# 0.0.4

- Add `markdownUnwrapElements` option
- Convert package to native ES module
- Will print debug statements to STDOUT if `DEBUG` env var is toggled
- Returns the destination image path instead of `undefined`

# 0.0.3

- Add `markdownTailwindClasses` option

# 0.0.2

- Add a `globals` option for having the calling function pass data to the
template, regardless of the markdown

# 0.0.1

- Initial release


lib/read-markdown.js
import matter from 'gray-matter'
import showdown from 'showdown'

export default async function readMarkdown (src, options) {
  const markdown = matter.read(src, { excerpt_separator: options.summarySeparator })

  const converter = new showdown.Converter({
    noHeaderId: true,
    extensions: [...extensionsFor(options.markdownTailwindClasses, options.markdownStyleAttrs, options.markdownUnwrapElements, options.markdownExtensions)]
  })
  return {
    data: markdown.data,
    content: {
      markdown: markdown.content,
      html: converter.makeHtml(markdown.content)
    },
    summary: {
      markdown: markdown.excerpt,
      html: converter.makeHtml(markdown.excerpt)
    },
    globals: options.globals
  }
}

// Copied this technique from
// https://github.com/showdownjs/showdown/wiki/Add-default-classes-for-each-HTML-element
function extensionsFor (twClassMap = {}, styleAttrsMap = {}, unwrapElements = [], extensions = []) {
  return unwrapElements.map(el => {
    return {
      type: 'output',
      regex: new RegExp(`<${el}[^>]*>([^<]*)<\\/${el}>`, 'g'),
      replace: '$1'
    }
  }).concat(Object.keys(twClassMap).map(el => {
    return {
      type: 'output',
      regex: new RegExp(`<${el}(.*)>`, 'g'),
      replace: `<${el} tw="${twClassMap[el]}"$1>`
    }
  }), Object.keys(styleAttrsMap).map(el => {
    return {
      type: 'output',
      regex: new RegExp(`<${el}(.*)>`, 'g'),
      replace: `<${el} style="${styleAttrsMap[el]}"$1>`
    }
  }), extensions)
}


index.js
import path from 'path'
import fs from 'fs'
import { createRequire } from 'node:module'
import { fileURLToPath } from 'url'

import log from './lib/log.js'
import readMarkdown from './lib/read-markdown.js'
import renderTemplate from './lib/render-template.js'
import convertSvgToPng from './lib/convert-svg-to-png.js'
import exportImage from './lib/export-image.js'

const __dirname = fileURLToPath(path.dirname(import.meta.url))
const require = createRequire(import.meta.url)

const DEFAULT_CONFIG = {
  template: path.join(__dirname, 'default-template.html'),
  width: 1200,
  height: 630,
  fonts: [
    {
      name: 'Source Sans Pro',
      data: fs.readFileSync(require.resolve('@fontsource/source-sans-pro/files/source-sans-pro-latin-400-normal.woff')),
      weight: 400,
      style: 'normal'
    }
  ]
}

export default async function realOg (src, dest, options) {
  log(`Writing ${dest} from ${src} markdown`)
  const page = await readMarkdown(src, { summarySeparator: '<!--more-->', ...options })
  log(`Rendering SVG for ${src}`)
  const svg = await renderTemplate(page, Object.assign({}, DEFAULT_CONFIG, options))
  log(`Converting SVG to PNG for ${src}`)
  const png = await convertSvgToPng(svg)
  log(`Exporting image for ${src}`)
  await exportImage(png, dest, { fileType: 'png', ...options })
  log(`Finished writing ${dest}`)
  return dest
}

/* This is not a function to be proud of.
 * It will render the markdown to SVG at a particular width and with the user's
 * supplied template and options, and then it'll run a regex against the
 * resulting SVG to find the height of the first rect in the output.
 *
 * The output currently starts like this, which the regex is based on :
 *  <svg width="2088" height="100000" viewBox="0 0 2088 100000" xmlns="http://www.w3.org/2000/svg"><mask id="satori_om-id"><rect x="0" y="0" width="2102" height="549" fill="#fff"/></mask>
 *
 * Obligatory you can't parse HTML with regex link: https://blog.codinghorror.com/parsing-html-the-cthulhu-way/
 */
export async function measureHeight (src, options) {
  const page = await readMarkdown(src, { summarySeparator: '<!--more-->', ...options })
  const svg = await renderTemplate(page, Object.assign({}, DEFAULT_CONFIG, options))
  return Number(svg.match(/<rect[^>]*height="(\d+)"/)?.[1])
}
